### ISSUE TITLE
Tricky cycles aren't handled

### ISSUE THREAD

**Issue by Rich-Harris** [rollup/rollup#36]
Cases like this...

```js
// a.js
import B from './B';

export default class A {
  constructor () {}

  b () {
    return new B();
  }
}

// b.js
import A from './A';

export default class B extends A {
  constructor () {
    super();
  }
}
```

...aren't dealt with very intelligently - if a statement that's already been included in the bundle is encountered (indicating a cyclical dependency), it isn't followed a second time. So in situations like the above, if we encounter `A` first, we include the definition for `B` (because it's used inside the `A` definition), which leads us back to `A` which is ignored, even though it needs to go first.

The solution with Esperanto was to distinguish between 'strong' and 'weak' dependencies (weak meaning inside a function'), and reordering bundles with cyclical dependencies to ensure that strong dependencies appeared at the top. We can actually improve on this - because of hoisting, nothing can ever strongly depend on a function. (EDIT: actually that's not true, you can assign properties to functions which means the functions must be declared first. A function can never strongly depend on anything else, but we get that for free anyway.)

**Comment by Rich-Harris** [comment]
[Failing test on gh-36 branch](https://github.com/rollup/rollup/tree/gh-36/test/function/cycles-pathological)

**Comment by Rich-Harris** [comment]
Fixed in 0.9.0
